[Keyboard] Add Pimoroni Keybow 2040

fanf2 opened the pull request:
## Description

![Keybow 2040](https://cdn.learn.pimoroni.com/article/assembling-keybow-2040/assets/keybow2040-03.jpg?width=800)

The [Pimoroni Keybow 2040][keybow2040] is a 16 key mechanical macropad controlled by a Raspberry Pi RP2040, with Kailh hot swap sockets and per key RGB lighting, in a decorative FR4 sandwich.

[keybow2040]: https://shop.pimoroni.com/products/keybow-2040

I am not associated with Pimoroni, I just own a Keybow 2040. I thought installing QMK on it might be a fun way to learn more about QMK. *LITTLE DID I KNOW...*

As far as I know, all the problems with this PR have been resolved, so it should be ready to go.

The following struck-out sections describe some problems I had when trying to get it to work. They are retained for historical interest only.

### VIA

~~On my M1 Mac, VIA.app is stuck "Searching for devices..."~~

~~I have tried loading `keybow2040.json` (from thte VIA commit in this PR) into VIA's "Design" tab: still no luck.~~

~~I ttried adding a little debug logging to `raw_hid_receive()` in `quantum/via.c`, but it never prints anything to the console.~~

~~I also have a Mechboards / Yiancar HS60 which _used_ to work with VIA, but does not any more.~~

~~So I currently suspect something is broken with VIA, but it does not seem possible to get any information out of VIA to explain what it is doing and what is or is not working, and searchengineering has so far not helped me find a solution.~~

### LEDs

~~Another frustrating failure, I'm afraid~~

~~I seem to be suffering from some kind of asynchronous corruption of the global `matrix[]` array. I tried to track it down with a lot of `VERIFY_TRACED_VARIABLES()` but I ended up at some code that could not possiblly be blamed, unless the corruption was coming from some kind of interrupt service routine.~~

### Despite that

I have tracked down the relevant source code for the Keybow 2040 default CircuitPython firmware, and together with its published schematic I think most of the configuration setttings are reasonably plausible (tho I have not been able to test things like the orientation of the LED matrix). The `readme.md` contains links to the more obscure information I used.

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project
    ~~*nah, I can fix that when the code work*~~ :trollface:
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation
   ~~*maybe, dunno yet if there are things I learned that could go into the docs*~~
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work
    ~~*heh lol not quite*~~ 😭

sigprof commented on keyboards/pimoroni/keybow2040/keybow2040.c:
```diff
@@ -0,0 +1,74 @@
+
+// https://github.com/adafruit/Adafruit_CircuitPython_IS31FL3731/blob/main/adafruit_is31fl3731/keybow2040.py
+
+const is31_led PROGMEM g_is31_leds[RGB_MATRIX_LED_COUNT] = {
+	{ 0, 120, 88, 104 }, // 0, 0
+	{ 0, 136, 40, 72 },  // 1, 0
+	{ 0, 112, 80, 96 },  // 2, 0
+	{ 0, 128, 32, 64 },  // 3, 0
+	{ 0, 121, 89, 105 }, // 0, 1
+	{ 0, 137, 41, 73 },  // 1, 1
+	{ 0, 113, 81, 97 },  // 2, 1
+	{ 0, 129, 33, 65 },  // 3, 1
+	{ 0, 122, 90, 106 }, // 0, 2
+	{ 0, 138, 25, 74 },  // 1, 2
+	{ 0, 114, 82, 98 },  // 2, 2
+	{ 0, 130, 17, 66 },  // 3, 2
+	{ 0, 123, 91, 107 }, // 0, 3
+	{ 0, 139, 26, 75 },  // 1, 3
+	{ 0, 115, 83, 99 },  // 2, 3
+	{ 0, 131, 18, 67 },  // 3, 3
+};
```
These numbers look wrong for the IS31FL3731 driver in QMK — that driver expects the register addresses which match this part of the IS31FL3731 datasheet (the PWM Register columns):
![IS31FL3731 datasheet](https://github.com/qmk/qmk_firmware/assets/616082/ff4a4a7f-ad6a-4922-92b9-2a54e25f9ede)

So the range of register addresses should be `0x24`…`0xB3` (36…179).  I guess that CircuitPython number the PWM registers starting from 0, so you need to add `0x24` to these numbers.  The next step would be replacing those raw numbers with the matching defines from [is31fl3731.h](https://github.com/qmk/qmk_firmware/blob/aa91d98531a299f86f9dab8875dcf46983b16af1/drivers/led/issi/is31fl3731.h#L50-L210) — those defines match the numbering in the datasheet.

fanf2 replied:
Excellent, thanks for spotting that mistake! I was too lazy to replace the raw numbers with the register names: regrettable, because I would have noticed the mismatch if I had cleaned them up.

fanf2 replied:
This is fixed in round two of the PR.

sigprof commented:
> On my M1 Mac, VIA.app is stuck "Searching for devices..."

Are you using [the latest VIA.app](https://github.com/the-via/releases/releases)?  Although it seems that the preferred way to use VIA now is to open https://usevia.app in a Chrome-based browser (the required WebHID support is currently not present in any other browser engine).  Any old version of the local VIA application won't work with recent firmwares (and new versions are just Electron wrappers for https://usevia.app).

sigprof commented on keyboards/pimoroni/keybow2040/info.json:
```diff
@@ -0,0 +1,59 @@
+        "vid": "0x16D0",
+        "pid": "0x08C6",
+    },
+    "eeprom": {
+	"driver": "transient",
+    },
```
The default EEPROM driver (which for RP2040 emulates EEPROM using a part of the same flash chip that contains the firmware) should work (at least the official schematic shows the W25Q16JVUXIQ chip, which should be supported).  Enabling the VIA support with the `transient` driver is mostly useless.
```suggestion
```

fanf2 replied:
Yes indeed :-) I left that in there as a debugging aid because I have been *very* confused by unexpected statefulness in the past! Sadly json does not let me explain things like this in the best place :-/

sigprof replied:
Technically `info.json` files are parsed as Hjson by the Python code and therefore allow comments and some other syntax extensions, but there is some disagreement about whether the usage of such things should actually be encouraged.

However, `keymap.json` files must be pure JSON (one reason is that QMK Configurator can fetch those files directly from the `qmk_firmware` repo in some cases).

fanf2 replied:
I've removed the debugging remnants from round two

sigprof commented on keyboards/pimoroni/keybow2040/config.h:
```diff
@@ -0,0 +1,20 @@
+
+#pragma once
+
+#define TRANSIENT_EEPROM_SIZE 1024
+
```
This should not be needed once you switch from `transient` to the default EEPROM driver.
```suggestion
```

fanf2 replied:
Gone in round two

sigprof commented on keyboards/pimoroni/keybow2040/config.h:
```diff
@@ -0,0 +1,20 @@
+// adafruit_is31fl3731 says 0x74 is the default
+#define DRIVER_ADDR_1 0b1110100
+
+#define I2C0_CLOCK_SPEED 50000
```
At the moment the RP2040 I2C driver is broken and does not produce proper timings, so you may need this value (it actually gives about 186 kHz in my testing).  Eventually it should be removed (the IS31FL3731 chip is supposed to support 400 kHz, but the current driver makes it about 706 kHz instead, so the chip might not work properly with the default speed configuration).

fanf2 replied:
Right. I saw a comment on another RP2040 keyboard QMK config saying that it needed a clock speed setting to work reliably. I didn’t know it was a bug in the driver, I was trying all sorts of things to see if they improved matters.

sigprof commented on keyboards/pimoroni/keybow2040/keybow2040.c:
```diff
@@ -0,0 +1,74 @@
+#include QMK_KEYBOARD_H
+#include "rgb_matrix.h"
+
+void keyboard_post_init_user(void) {
+	debug_enable = true;
+}
+
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+#ifdef CONSOLE_ENABLE
+	dprintf("KL: kc: 0x%04X, col: %2u, row: %2u, "
+		"pressed: %u, time: %5u, "
+		"int: %u, count: %u\n",
+		keycode, record->event.key.col, record->event.key.row,
+		record->event.pressed, record->event.time,
+		record->tap.interrupted, record->tap.count);
+#endif
+  return true;
+}
+
```
The `*_user` functions must not be defined at the keyboard level, and this debugging code probably should be removed from the final version anyway.
```suggestion
```

fanf2 replied:
Absolutely: this is like the `eeprom_transient` setting.

I realise I forgot to mention in the PR description that I deliberately left in some debugging setup. (Note to self: mentioning it in the commit messages is not enough.)  Oops, sorry for wasting your time!

sigprof replied:
I don't even try to read commit messages in PRs to `qmk_firmware`, because at the merge time the PR will get squashed to a single commit anyway, and therefore there is no requirement to keep the individual commit messages in PRs clean.

fanf2 replied:
Gone in round two; I've pre-squashed the PR to one commit as well

fanf2 commented:
Yes, I installed VIA 2.2 and 3.0, neither worked.

I prefer to avoid Google as much as possible. I wonder if VIA is failing at rawhid because of Apple security restrictions on an unsigned executable, but who can say without debug logging 🤷  Maybe the Chrome dev tools will be more helpful. Yuck.

sigprof commented on keyboards/pimoroni/keybow2040/keybow2040.c:
```diff
@@ -0,0 +1,74 @@
+	{ 0, 131, 18, 67 },  // 3, 3
+};
+
+led_config_t g_led_config = {
+	// Matrix location to LED index
+	{
+		{ 0x0, 0x1, 0x2, 0x3, },
+		{ 0x4, 0x5, 0x6, 0x7, },
+		{ 0x8, 0x9, 0xA, 0xB, },
+		{ 0xC, 0xD, 0xE, 0xF, },
+		{ NO_LED, NO_LED, NO_LED, NO_LED, },
+	},
+
+	// LED index to physical location
+	// equally spaced horizontally;
+	// vertically LEDs are 5mm from the top edge
+	{
+		{25,7}, {75,7}, {125,7}, {175,7},
+		{25,57}, {75,57}, {125,57}, {175,57},
+		{25,107}, {75,107}, {125,107}, {175,107},
+		{25,157}, {75,157}, {125,157}, {175,157},
+	},
+
+	// LED index to flags
+	{
+		LED_FLAG_KEYLIGHT, LED_FLAG_KEYLIGHT,
+		LED_FLAG_KEYLIGHT, LED_FLAG_KEYLIGHT,
+		LED_FLAG_KEYLIGHT, LED_FLAG_KEYLIGHT,
+		LED_FLAG_KEYLIGHT, LED_FLAG_KEYLIGHT,
+		LED_FLAG_KEYLIGHT, LED_FLAG_KEYLIGHT,
+		LED_FLAG_KEYLIGHT, LED_FLAG_KEYLIGHT,
+		LED_FLAG_KEYLIGHT, LED_FLAG_KEYLIGHT,
+		LED_FLAG_KEYLIGHT, LED_FLAG_KEYLIGHT,
+	}
+};
```
The current preference is to convert this data into `info.json`.
`qmk info -f json -kb pimoroni/keybow2040` should be able to make the JSON data for `rgb_matrix` from this.

drashna replied:
`qmk migrate -kb pimoroni/keybow2040` may be simpler, as it should move most of the settings it can into the json and remove the legacy config.

fanf2 replied:
Migrated to info.json in round two

@drashna I tried `qmk migrate` but that only migrated the `center_point`. `qmk info -f json` was able to turn the `g_led_config` into JSON `rgb_matrix` settings, so I used that. I was surprised that `qmk info -f json` knows more than `qmk migrate`

sigprof commented on keyboards/pimoroni/keybow2040/keymaps/default/keymap.json:
```diff
@@ -0,0 +1,23 @@
+    "version": 1,
+    "keymap": "default",
+    "layout": "LAYOUT",
+    "layers": [
+	[
+            "RGB_TOG", "RGB_MOD", "KC_C",    "KC_D",
+            "KC_E",    "KC_F",    "KC_G",    "KC_H",
+            "KC_I",    "KC_J",    "KC_K",    "KC_L",
+            "KC_M",    "KC_N",    "KC_O",    "KC_P",
+	    "KC_Q"
+	],
+	[
+            "KC_F1",    "KC_F2",    "KC_F3",    "KC_F4",
+            "KC_F5",    "KC_F6",    "KC_F7",    "KC_F8",
+            "KC_F9",    "KC_F10",   "KC_F11",   "KC_F12",
+            "KC_F13",   "KC_F14",   "KC_F15",   "KC_F16",
+	    "KC_F17"
+	]
```
The second layer is mostly useless, because you don't have any way to switch to that layer.

fanf2 replied:
Heh, yes :-)

I am not sure what the best default for a macro pad should be: function keys seem benign and not totally useless for plug-and-play beginners, but I guess the most versatile config would be a load of dynamic macro keys that VIA can edit.

When I was using it to extend a HHKB I had F keys in the top half and a navigation cluster in the bottom half. The Pimoroni CircuitPython firmware needs configuration with Python to do anything.

fanf2 replied:
I have settled on a defensible default layout in round 2

sigprof commented on keyboards/pimoroni/keybow2040/keymaps/via/keybow2040.json:
This file needs to be removed from the QMK PR (it should be submitted to https://github.com/the-via/keyboards/ instead).

And once you fix RGB Matrix, you would be able to add the VIA support for that as described in https://www.caniusevia.com/docs/v3_changes (any new boards need to use the V3 format anyway).

fanf2 replied:
Gone in round twto

sigprof commented on keyboards/pimoroni/keybow2040/keymaps/via/keymap.json:
```diff
@@ -0,0 +1,42 @@
+            "KC_M",    "KC_N",    "KC_O",    "KC_P",
+	    "KC_Q"
+	],
+	[
+            "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS",
+            "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS",
+            "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS",
+            "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS",
+            "KC_TRNS"
+	],
+	[
+            "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS",
+            "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS",
+            "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS",
+            "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS",
+            "KC_TRNS"
+	],
+	[
+            "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS",
+            "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS",
+            "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS",
+            "KC_TRNS", "KC_TRNS", "KC_TRNS", "KC_TRNS",
+            "KC_TRNS"
+	]
```
Padding the VIA keymap to the maximum number of layers is no longer needed — recent QMK can detect the number of actually initialized layers and fill any uninitialized layers with `KC_TRNS` automatically.
```suggestion
```

fanf2 replied:
Cleaned up in round two

sigprof commented on keyboards/pimoroni/keybow2040/readme.md:
```diff
@@ -0,0 +1,83 @@
+
+  * The BOOTSEL button doubles as a 17th configurable key.
+
+  * There is no USB PID allocated for the Keybow 2040 under
+    [the Raspberry Pi USB VID](https://github.com/raspberrypi/usb-pid),
+    but its CircuitPython port uses VID = 0x16D0, PID = 0x08C6
+    so I copied that.
```
I'm not sure whether copying the USB ID from CircuitPython is the proper thing to do — this may create some conflicts, like VIA trying to poke a board that uses the CircuitPython firmware and therefore doing some unexpected things.

fanf2 replied:
Oh good, I thought it might be a bad idea too. (Also the CircuitPython VID is a real VID and the PID only fails to collide by luck.) I will go back to my original plan of cooking up a new “random” pair of IDs.

sigprof replied:
0x16D0 is the VID for which unique PID values are sold by [MCS Electronics](https://www.mcselec.com/index.php?page=shop.product_details&product_id=92&option=com_phpshop), so the PID can probably be considered as “owned” by Pimoroni.

The problem with VIA is that it does not perform a proper detection for the HID usage page to find the rawhid interface (although that usage page is in the vendor range, so in theory it might collide with some completely unrelated usage).  Another potential problem is that Windows in some circumstances can cache the binding of drivers to USB interfaces, therefore you may have some trouble if the device changes the set of advertised interfaces without changing its USB VID:PID pair.

fanf2 replied:
Oh, I didn’t know MCS did that, cool. And thanks for those details about why it is best to avoid this kind of collision.

fanf2 replied:
I picked new USB IDs in round two, and I checked [it does not collide](https://devicehunt.com/search/type/pci/vendor/07F8/device/any)

sigprof commented:
> I prefer to avoid Google as much as possible.

On Linux [ungoogled-chromium](https://github.com/ungoogled-software/ungoogled-chromium) can be used with https://usevia.app; no idea whether it would work on MacOS (especially on the M1 platform).

drashna commented:
> > I prefer to avoid Google as much as possible.
> 
> On Linux [ungoogled-chromium](https://github.com/ungoogled-software/ungoogled-chromium) can be used with [usevia.app](https://usevia.app); no idea whether it would work on MacOS (especially on the M1 platform).

For Ventura, I found that I had to explicitly add the via app to the "input monitoring" permissions before via would work.  Same with the toolbox for the console/hid bootloaders.

fanf2 commented:
I have pushed round two of this pull request 🎉 

Many thanks to @sigprof for the thorough and helpful review, and thanks to @drashna for the extra hintts and tips!

I have gone through the review comments and I think I have addressed them all. My apologies if I missed any.

I got the RGB matrix working after much careful checking of each LED's register number, and the corresponding key and LED locations should match. (I understand now how the incorrect LED register numbers could cause out-of-bounds array accesses which would explain the mysterious memory corruption I was struggling with before.)

There were also some problems with inconsistencies in my I2C driver configuration. I cleaned them up to match the board's schematic: it uses GPIOs 4 and 5 for I2C, and those pins are connected to the RP2040's I2C0 peripheral. However, a perfectly consistent I2C0 configuration does not compile because some of QMK's code assumes that I2C1 is configured. If I add a dummy I2C1 configuration that refers to unused RP2040 pins, the RGB matrix remains black. If I tell the driver to use the same GPIO pins for I2C0 and I2C1, it compiles and works. 🧐 

I have not yet tested that VIA works if I use Chrome. Maybe after I have had something to eat.

sigprof commented on keyboards/pimoroni/keybow2040/mcuconf.h:
```diff
@@ -0,0 +1,37 @@
+
+#include_next <mcuconf.h>
+
+/*
+ * implicitly LED driver chips
+ */
+#define DRIVER_COUNT 1
+
+/*
+ * I2C address when IS31FL3731 AD pin is connected to ground
+ */
+#define DRIVER_ADDR_1 0b1110100
+
```
`DRIVER_COUNT` and `DRIVER_ADDR_1` are QMK configuration parameters; they should be set in `config.h`, not in `mcuconf.h`.
```suggestion
```

fanf2 replied:
Fixed in round three!

sigprof commented on keyboards/pimoroni/keybow2040/mcuconf.h:
```diff
@@ -0,0 +1,37 @@
+#undef  RP_I2C_USE_I2C1
+#define RP_I2C_USE_I2C0 TRUE
+#define RP_I2C_USE_I2C1 FALSE
+
+#undef  I2C_DRIVER
+#define I2C_DRIVER I2CD0
+
+#define I2C0_SDA_PIN GP4
+#define I2C0_SCL_PIN GP5
+#define I2C0_CLOCK_SPEED 100000
+
+/*
+ * I2C1 should be unused, but there seem to be parts of the driver
+ * that assume I2C1 is used and I2C0 is not, so tell both of them to
+ * use the same GPIO pins. This is necessary for it to build and work
+ * correctly.
+ */
+#define I2C1_SDA_PIN GP4
+#define I2C1_SCL_PIN GP5
+#define I2C1_CLOCK_SPEED 100000
```
These parameters should also be set in `config.h`, not here.
```suggestion
```

fanf2 replied:
Fixed in round three!

sigprof commented on keyboards/pimoroni/keybow2040/config.h:
```diff
@@ -0,0 +1,7 @@
+
+#define RP2040_BOOTLOADER_DOUBLE_TAP_RESET
+
+#define RGB_MATRIX_LED_COUNT 16
```
```suggestion
// I2C config
#define I2C_DRIVER I2CD0
#define I2C1_SDA_PIN GP4
#define I2C1_SCL_PIN GP5
#define I2C1_CLOCK_SPEED 100000

// The board has a single IS31FL3731 chip with the AD pin connected to GND.
#define DRIVER_COUNT 1
#define DRIVER_ADDR_1 0b1110100

#define RGB_MATRIX_LED_COUNT 16
```
I2C configuration options can look confusing, because for historical reasons the names of some defines include the `1` index.  In these names (`I2C1_SDA_PIN`, `I2C1_SCL_PIN`, `I2C1_CLOCK_SPEED`) the `1` index does not refer to the I2C peripheral number of the MCU — it's something like “the index of the I2C bus on the keyboard PCB”, except the QMK code does not actually support more than a single I2C bus yet, therefore the only valid value of that index is 1.  The only place where the actual I2C peripheral number is specified is the value of the `I2C_DRIVER` define — `I2CD0` refers to the `I2C0` peripheral as described in the MCU datasheet.

fanf2 replied:
Ohhhhh 💡 that makes a lot more sense than my less charitable assumption!

sigprof commented on keyboards/pimoroni/keybow2040/info.json:
```diff
@@ -0,0 +1,116 @@
+    },
+    "usb": {
+        "device_version": "1.0.0",
+        "vid": "0x07F8",
+        "pid": "0x4784",
```
Unfortunately, that `0x07F8` value is too low — it's likely to be assigned to some company, but USB-IF does not make these assignments public, therefore any public databases of USB VIDs are definitely incomplete.  Currently the way to handle USB VID:PID values in QMK is mostly “pick some random number above `0x4000` for VID, check for collisions in any public databases and https://yanfali.github.io/qmk_usb_usage/, then just hope for the best”.

fanf2 replied:
Also fixed in round three

sigprof commented on keyboards/pimoroni/keybow2040/info.json:
```diff
@@ -0,0 +1,116 @@
+        "vid": "0x07F8",
+        "pid": "0x4784",
+    },
+    "rgb_matrix": {
+        "driver": "IS31FL3731",
+	"animations": {
+	    "alpha_mods": true,
```
Apparently this file contains tab characters, so it looks unaligned because of some tab size mismatch.  Applying `qmk format-json` to this file should fix the formatting.

fanf2 replied:
I untabified all the json in round three

sigprof commented on keyboards/pimoroni/keybow2040/halconf.h:
```diff
@@ -0,0 +1,8 @@
+
+#pragma once
+
+#include_next <halconf.h>
+
+#undef  HAL_USE_I2C
+#define HAL_USE_I2C TRUE
```
In `halconf.h` the order of `#define` and `#include_next` should be different (ChibiOS is not consistent here, you just need to remember that for `halconf.h` the defines go first, and for `mcuconf.h` the undef/define pairs go last).
```suggestion
#define HAL_USE_I2C TRUE

#include_next <halconf.h>
```

fanf2 replied:
Heh, fixed in round three

fanf2 commented:
Round three!

Thanks again to @sigprof for your helpful guidance. I think I've addressed the problems you found. I have added some improved explanatory comments to `keybow2040/config.h`; I hope they are correct!

mosch commented:
@fanf2 thanks for working on this! I can confirm the fw works on my Keybow 2040. Im also having issues with the via app. Tested the downloadable version and [usevia.app](https://usevia.app/);  on Win+Mac in Chrome + Edge. @fanf2 how did you get it working? 
The Keybow shows up in the browser dialog, and shows as paired. However I only see the [Authorized Device +] Button.

fanf2 commented:
Hi, @mosch,

VIA does not yet know about the Keybow 2040; I have been waiting for this PR to be merged before I open the PR for VIA supprt.

Until it gets official support from VIA, you need to make the following incantations:

  * go to the settings (cogwheel) tab in VIA;
  * turn on the "show design tab" switch
  * go to the design (paintbrush) tab;
  * bop the "load draft definition" button
  * feed it a copy of the following JSON

```
{
    "name": "Pimoroni Keybow 2040",
    "vendorId": "0x5069",
    "productId": "0x4784",
    "matrix": {
	"rows": 5,
	"cols": 4
    },
    "layouts": {
	"keymap": [
	    [{"x":2.5,"w":0.5,"h":0.5,"w2":1},"4,0"],
	    [{"y":-0.5},"0,0","0,1","0,2","0,3"],
	    ["1,0","1,1","1,2","1,3"],
	    ["2,0","2,1","2,2","2,3"],
	    ["3,0","3,1","3,2","3,3"]
	]
    }
}
```

mosch commented:
Sweet that works, Thank you!

mosch commented on keyboards/pimoroni/keybow2040/info.json:
```diff
@@ -0,0 +1,116 @@
+    "bootloader": "rp2040",
+    "board": "GENERIC_RP_RP2040",
+    "features": {
+        "console": true
```
```suggestion
        "console": true,
```

fanf2 replied:
Thanks for spotting those mistakes! New revision incoming!

mosch reviewed (commented):
I think we are missing a comma here. I also found that this JSON uses tailing commas, while the others don't.

mosch commented:
This looks good now, anything missing before getting it in? @sigprof

drashna commented on keyboards/pimoroni/keybow2040/keybow2040.c:
```diff
@@ -0,0 +1,27 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include QMK_KEYBOARD_H
+#include "rgb_matrix.h"
```
```suggestion
#include "quantum.h"
```

drashna commented on keyboards/pimoroni/keybow2040/keybow2040.c:
```diff
@@ -0,0 +1,27 @@
+/*
+ * see Keybow 2040 schematic for the LED wiring
+ */
+const is31_led PROGMEM g_is31_leds[RGB_MATRIX_LED_COUNT] = {
+    /* driver, r, g, b */
+    { 0, C8_9,  C6_9,  C7_9  }, // LEDB12 0,0
+    { 0, C9_9,  C3_9,  C5_9  }, // LEDB11 1,0
+    { 0, C8_1,  C6_1,  C7_1  }, // LEDA12 2,0
+    { 0, C9_1,  C3_1,  C5_1  }, // LEDA11 3,0
+    { 0, C8_10, C6_10, C7_10 }, // LEDB22 0,1
+    { 0, C9_10, C3_10, C5_10 }, // LEDB21 1,1
+    { 0, C8_2,  C6_2,  C7_2  }, // LEDA22 2,1
+    { 0, C9_2,  C3_2,  C5_2  }, // LEDA21 3,1
+    { 0, C8_11, C6_11, C7_11 }, // LEDB32 0,2
+    { 0, C9_11, C2_10, C5_11 }, // LEDB31 1,2
+    { 0, C8_3,  C6_3,  C7_3  }, // LEDA32 2,2
+    { 0, C9_3,  C2_2,  C5_3  }, // LEDA31 3,2
+    { 0, C8_12, C6_12, C7_12 }, // LEDB42 0,3
+    { 0, C9_12, C2_11, C5_12 }, // LEDB41 1,3
+    { 0, C8_4,  C6_4,  C7_4  }, // LEDA42 2,3
+    { 0, C9_4,  C2_3,  C5_4  }, // LEDA41 3,3
+};
```
```suggestion
#ifdef RGB_MATRIX_ENABLE
const is31_led PROGMEM g_is31_leds[RGB_MATRIX_LED_COUNT] = {
    /* driver, r, g, b */
    { 0, C8_9,  C6_9,  C7_9  }, // LEDB12 0,0
    { 0, C9_9,  C3_9,  C5_9  }, // LEDB11 1,0
    { 0, C8_1,  C6_1,  C7_1  }, // LEDA12 2,0
    { 0, C9_1,  C3_1,  C5_1  }, // LEDA11 3,0
    { 0, C8_10, C6_10, C7_10 }, // LEDB22 0,1
    { 0, C9_10, C3_10, C5_10 }, // LEDB21 1,1
    { 0, C8_2,  C6_2,  C7_2  }, // LEDA22 2,1
    { 0, C9_2,  C3_2,  C5_2  }, // LEDA21 3,1
    { 0, C8_11, C6_11, C7_11 }, // LEDB32 0,2
    { 0, C9_11, C2_10, C5_11 }, // LEDB31 1,2
    { 0, C8_3,  C6_3,  C7_3  }, // LEDA32 2,2
    { 0, C9_3,  C2_2,  C5_3  }, // LEDA31 3,2
    { 0, C8_12, C6_12, C7_12 }, // LEDB42 0,3
    { 0, C9_12, C2_11, C5_12 }, // LEDB41 1,3
    { 0, C8_4,  C6_4,  C7_4  }, // LEDA42 2,3
    { 0, C9_4,  C2_3,  C5_4  }, // LEDA41 3,3
};
#endif
```

itsnelsonmartins commented on keyboards/pimoroni/keybow2040/info.json:
```diff
@@ -0,0 +1,116 @@
+        "pid": "0x4784"
+    },
+    "rgb_matrix": {
+        "driver": "IS31FL3731",
```
change from IS31FL3731 to is31fl3731

drashna replied:
```suggestion
        "driver": "is31fl3731",
```

fanf2 replied:
This works after I rebase on top of current master

itsnelsonmartins reviewed (changes requested):
change from IS31FL3731 to is31fl3731 to fix QMK CI Run

fanf2 commented:
OK, I have updated the `readme.md` to follow the template more closely. I don't have an imgur account so I deleted the hero image. I have improved `keybow2040.c`, `config.h`, and `info.json` as suggested.

Thank you for the review and style fixes. I hope it is now good to go!

drashna commented on keyboards/pimoroni/keybow2040/readme.md:
```diff
@@ -0,0 +1,101 @@
+Keybow 2040
+===========
```
```suggestion
# Keybow 2040
```

drashna commented on keyboards/pimoroni/keybow2040/readme.md:
```diff
@@ -0,0 +1,101 @@
+[Complete Newbs Guide](https://docs.qmk.fm/#/newbs).
+
+
+bootloader
+----------
+
+The procedure to flash the firmware is similar to other RP2040
+devices. You can enter the bootloader in two ways:
+
+  * **Vulcan nerve pinch**: hold down BOOTSEL while pressing RESET
+    (not so great if you have configured BOOTSEL to do something)
+
+  * **Mr Resetti double-tap**: press RESET twice quickly
+    (only works when running the QMK firmware)
+
+Find the BOOTSEL and RESET buttons on your Keybow 2040 either side of
+the USB socket on the edge of the PCB. When the USB socket is facing
+away from you, BOOTSEL is on the right and RESET is on the left.
+According to the schematic the reset button is called Mr Resetti.
+
+  * _Aside_: The BOOTSEL button disables the QSPI flash that stores
+    the firmware, which makes the RP2040 use its built-in UF2
+    bootloader.
+
+A USB drive will appear called `RPI-RP2`. Copy the firmware file
+`pimoroni_keybow2040_default.uf2` onto the `RPI-RP2` drive using
+whatever software is convenient (e.g. the `cp` command, the macOS
+Finder, or Windows Explorer). The `make :flash` command will copy the
+firmware for you.
+
+The `RPI-RP2` drive will disappear (which might make your OS complain
+that you failed to unmount it first) and a keyboard device will appear
+in its place.
+
+The Pimoroni web site has [instructions for reinstalling CircuitPython
+on the Keybow 2040](https://learn.pimoroni.com/article/circuitpython-and-keybow-2040)
+
```
```suggestion
## Bootloader 

Enter the bootloader in 3 ways:

* **Bootmagic reset**: Hold down the key at (0,0) in the matrix (usually the top left key or Escape) and plug in the keyboard
* **Physical reset button**: Double tap the RESET button, or hold BOOTSEL and tap RESET

The Pimoroni web site has [instructions for reinstalling CircuitPython
on the Keybow 2040](https://learn.pimoroni.com/article/circuitpython-and-keybow-2040)
```

fanf2 replied:
This change is wrong, because I have not enabled the bootmagic feature:

  * there's no point since the reset and bootsel buttons are easily accessible on the side of the Keybow2040
  * the RP2040's ROM implements the bootsel feature, so there's no need for QMK to do so
  * the bootsel button is not (0,0) in the matrix because that would mess up the 4x4 layout

I've trimmed the tutorial parts of that section, and added a sentence about bootsel vs bootmagic.

I've left my jokes in because this is supposed to be fun 🤪

drashna replied:
Bootmagic doesn't just reset the board and jump to bootloader, it also clears the eeprom.  Which can be very useful.  And has nothing to do with the bootsel button (it's just an arbitrary spot in the matrix, but defaults to 0,0).

drashna commented on keyboards/pimoroni/keybow2040/readme.md:
```diff
@@ -0,0 +1,101 @@
+on the Keybow 2040](https://learn.pimoroni.com/article/circuitpython-and-keybow-2040)
+
+
+notes
+-----
+
+  * The BOOTSEL button doubles as a 17th configurable key.
+
+  * The default keymap has F13 - F20 on the top half of the board
+    (nearest the USB socket), and a navigation cluster (arrows, home,
+    end, page up/down) on the bottom half. The BOOTSEL button changes
+    the RGB matrix mode.
+
+  * The Keybow 2040 CircuitPython firmware uses VID = 0x16D0, PID = 0x08C6,
+    but this QMK firmware has different USB IDs to avoid confusing drivers:
+    VID = 0x5069 ('Pi'), PID = 0x4784 (4 'x' 4)
+
```
Not sure how useful this is here. 
```suggestion
```

fanf2 replied:
I've left that in because there ought to be somewhere for users to find out what the default keymap is without reading the JSON, and I need somewhere to explain the info.json configuration choices. JSON isn't supposed to have comments, so 🤷

drashna commented on keyboards/pimoroni/keybow2040/readme.md:
```diff
@@ -0,0 +1,101 @@
+    VID = 0x5069 ('Pi'), PID = 0x4784 (4 'x' 4)
+
+
+resources
+---------
+
+Here are some links to the information I used when configuring QMK for the Keybow 2040.
+
+  * [Piomoroni Keybow 2040 product page](https://shop.pimoroni.com/products/keybow-2040)
+  * [Keybow 2040 assembly instructions](https://learn.pimoroni.com/article/assembling-keybow-2040)
+
+Places to find information about the Keybow 2040 hardware:
+
+  * [schematic](https://cdn.shopify.com/s/files/1/0174/1800/files/keybow_2040_schematic.pdf)
+      * contains most of the information that can also be gleaned from the CircuitPython support
+
+  * Pimoroni user-friendly CircuitPython libraries
+      * [old](https://github.com/pimoroni/keybow2040-circuitpython) and [new](https://github.com/pimoroni/pmk-circuitpython)
+      * these are wrappers around the low-level support below
+
+  * [CircuitPython hardware support](https://github.com/adafruit/circuitpython/blob/main/ports/raspberrypi/boards/pimoroni_keybow2040/)
+      * defines map from hardware feature names (e.g. `SW15`) to the way they are accessed (e.g. GPIO6)
+
+  * [CircuitPython IS31FL3731 driver](https://github.com/adafruit/Adafruit_CircuitPython_IS31FL3731/blob/main/adafruit_is31fl3731/keybow2040.py)
+      * defines the RGB matrix layout
```
This would be better linked in the "hardware supported" section above. Though, it looks like most of this is linked in the product page, already. 
```suggestion
```

fanf2 replied:
I have left these links in because it was difficult to find the relevant parts of the default Python firmware, so I think it will be helpful when anyone else wants to work on this board.

drashna replied:
But this isn't the circuit python repo, so has zero relevance here.  And most of the links add no value.

mosch commented:
How can I help getting this in? I think it's a vaulable contribution and brings a great keyboard into QMK. Let's get this in!

willoucom commented:
Hello, 
I have compiled and tested this firmware and it works perfectly.
Is it possible to review this PR again?
Can I help?
Thank you very much

keyboard-magpie commented:
@fanf2 will need to accept some of the suggested changes above still.

Additionally, there are some changes required since a round of breaking changes has passed- `qmk migrate` and/or `qmk info -kb pimoroni/keybow2040 -f json` may help you find these.

zvecr commented on keyboards/pimoroni/keybow2040/config.h:
```diff
@@ -0,0 +1,32 @@
+ */
+#define DRIVER_COUNT 1
+
+/*
+ * I2C address: IS31FL3731 AD pin is connected to ground
+ */
+#define DRIVER_ADDR_1 0b1110100
+
+/*
+ * The RP2040 has two I2C peripherals; I2C0 must be used with GPIO 4
+ * and 5 which on the Keybow 2040 are wired up to the IS31FL3731.
+ */
+#define I2C_DRIVER I2CD0
+
+/*
+ * I2C1 here refers to the board's / QMK's first (and only) I2C
+ * peripheral, not to the RP2040's second I2C peripheral.
+ */
```
```suggestion
#define IS31FL3731_I2C_ADDRESS_1 IS31FL3731_I2C_ADDRESS_GND

#define I2C_DRIVER I2CD0
```

fauxpark replied:
ISSI addresses should now use their respective defines instead of literals, for consistency and clarity.

zvecr replied:
Because things were refactored last develop cycle?

fanf2 replied:
OK, I see `#define DRIVER_ADDR_1` needs changing since d56ee70c524b4fc4d1638e5e8c4bdeb89e752993

Other than that, it's wrong to remove the comments since they explain where the values come from.

fanf2 commented on keyboards/pimoroni/keybow2040/rules.mk:
```diff
@@ -0,0 +1 @@
+# Blankety Blank chequebook and pen
```
Good grief am I not allowed to express my personality at all?

zvecr replied:
No, as it makes refactoring at a later date more complicated.

fanf2 commented on keyboards/pimoroni/keybow2040/halconf.h:
```diff
@@ -0,0 +1,10 @@
+
+#pragma once
+
+/*
+ * see config.h for notes on the Keybow 2040 I2C setup
+ */
```
I'm not removing this because it's an important cross-reference that I wanted (but lacked) when I was working on this code.

fauxpark replied:
config.h is a standard file in QMK. There is absolutely no need to tell users about it here.

drashna replied:
Not sure how adding that adds any value over what is in the docs. 
 RE: https://docs.qmk.fm/#/i2c_driver?id=arm-configuration

fanf2 commented on keyboards/pimoroni/keybow2040/mcuconf.h:
```diff
@@ -0,0 +1,13 @@
+
+#include_next <mcuconf.h>
+
+/*
+ * see config.h for notes on the Keybow 2040 I2C setup
+ */
+#undef  RP_I2C_USE_I2C0
+#undef  RP_I2C_USE_I2C1
+#define RP_I2C_USE_I2C0 TRUE
+#define RP_I2C_USE_I2C1 FALSE
```
Nah, I prefer my version.

fauxpark replied:
You can close this PR then if you like.

fanf2 commented on keyboards/pimoroni/keybow2040/config.h:
```diff
@@ -0,0 +1,32 @@
+#pragma once
+
+#define RP2040_BOOTLOADER_DOUBLE_TAP_RESET
+
+#define RGB_MATRIX_LED_COUNT (4 * 4)
+#define RGB_MATRIX_KEYPRESSES
+
+/*
+ * LED driver chip count: one IS31FL3731
+ */
+#define DRIVER_COUNT 1
+
```
Dunno why this change is needed?

fauxpark replied:
All three of these defines are no longer necessary.

zvecr replied:
Because they are no longer required.

fanf2 replied:
OK, I guess the DRIVER_COUNT is obsolete since 84df69572ce47f74213869120211f9db371cd3cf

And RGB_MATRIX_LED_COUNT is obsolete since 95681b8ff4a92aacd0249e124d34cf16e510175e

And DRIVER_COUNT since 84df69572ce47f74213869120211f9db371cd3cf

Dunno what the issue is with RGB_MATRIX_KEYPRESSES?

fauxpark replied:
It's implied when you have reactive effects enabled.

fanf2 commented on keyboards/pimoroni/keybow2040/keybow2040.c:
```diff
@@ -0,0 +1,28 @@
+#include "quantum.h"
+
+#ifdef RGB_MATRIX_ENABLE
+/*
+ * see Keybow 2040 schematic for the LED wiring
+ */
```
I'm not removing this because it's an important cross-reference that I wanted (but lacked) when I was working on this code.

fanf2 commented:
OK, I've gone through the comments and worked out where most of the unexplained ones came from, so I think this PR is now up-to-date with the upstream changes in the last three months.

I have reflashed my keybow2040 and it seems to be working ok.

fauxpark commented on keyboards/pimoroni/keybow2040/keybow2040.c:
```diff
@@ -0,0 +1,28 @@
+/*
+ * the LED addresses must match the wiring in the Keybow 2040 schematic
+ */
+const is31_led PROGMEM g_is31_leds[RGB_MATRIX_LED_COUNT] = {
```
```suggestion
const is31fl3731_led_t PROGMEM g_is31fl3731_leds[IS31FL3731_LED_COUNT] = {
```

fauxpark commented on keyboards/pimoroni/keybow2040/info.json:
```diff
@@ -0,0 +1,116 @@
+            ["GP19", "GP15", "GP11", "GP7"],
+            ["GP20", "GP16", "GP12", "GP8"],
+            ["GP21", "GP17", "GP13", "GP9"],
+            ["GP23", "NO_PIN", "NO_PIN", "NO_PIN"]
```
```suggestion
            ["GP23", null, null, null]
```

Gadgetoid commented:
@fanf2 thank you for all the effort you've put in here.

Could I please implore you to defer to the maintainers and get this PR ship-shaped when you have the time?

I would love to see this PR merged and for us (disclaimer: I'm the software wrangler at Pimoroni) to have a toe into the QMK world.

Thank you! :tada:
